gitlab/resource_gitlab_project_mirror.go

```diff
@@ -0,0 +1,176 @@
+
+	mirrorID := strconv.Itoa(mirror.ID)
+	d.SetId(buildTwoPartID(&projectID, &mirrorID))
+	return resourceGitlabProjectMirrorRead(d, meta)
+}
+
+func resourceGitlabMirrorUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+
+	mirrorID := d.Get("mirror_id").(int)
+	projectID := d.Get("project").(string)
+	enabled := d.Get("enabled").(bool)
+	onlyProtectedBranches := d.Get("only_protected_branches").(bool)
+	keepDivergentRefs := d.Get("keep_divergent_refs").(bool)
+
+	options := gitlab.EditProjectMirrorOptions{
+		Enabled:               &enabled,
+		OnlyProtectedBranches: &onlyProtectedBranches,
+		KeepDivergentRefs:     &keepDivergentRefs,
+	}
+	log.Printf("[DEBUG] update gitlab project mirror %v for %s", mirrorID, projectID)
+
+	_, _, err := client.ProjectMirrors.EditProjectMirror(projectID, mirrorID, &options)
+	if err != nil {
+		return err
+	}
+	return resourceGitlabProjectMirrorRead(d, meta)
+}
+
+// Documented remote mirrors API does not support a delete method, instead mirror is disabled.
+func resourceGitlabMirrorDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+
+	enabled := false
+
+	mirrorID := d.Get("mirror_id").(int)
+	projectID := d.Get("project").(string)
+	onlyProtectedBranches := d.Get("only_protected_branches").(bool)
+	keepDivergentRefs := d.Get("keep_divergent_refs").(bool)
+
+	options := gitlab.EditProjectMirrorOptions{
+		Enabled:               &enabled,
+		OnlyProtectedBranches: &onlyProtectedBranches,
+		KeepDivergentRefs:     &keepDivergentRefs,
+	}
+	log.Printf("[DEBUG] Disable gitlab project mirror %v for %s", mirrorID, projectID)
+
+	_, _, err := client.ProjectMirrors.EditProjectMirror(projectID, mirrorID, &options)
+
+	return err
+}
+
+func resourceGitlabProjectMirrorRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+
+	ids := strings.Split(d.Id(), ":")
+	projectID := ids[0]
+	mirrorID := ids[1]
+	integerMirrorID, err := strconv.Atoi(mirrorID)
+	if err != nil {
+		return err
+	}
+	log.Printf("[DEBUG] read gitlab project mirror %s id %v", projectID, mirrorID)
+
+	mirrors, _, err := client.ProjectMirrors.ListProjectMirror(projectID)
+
+	if err != nil {
+		return err
+	}
+
+	var mirror *gitlab.ProjectMirror
+	found := false
+
+	for _, m := range mirrors {
+		log.Printf("[DEBUG] project mirror found %v", m.ID)
+		if m.ID == integerMirrorID {
+			mirror = m
+			found = true
+		}
+	}
+
+	if !found {
+		d.SetId("")
+		return nil
+	}
+
+	resourceGitlabProjectMirrorSetToState(d, mirror, &projectID)
+	return nil
+}
+
+func resourceGitlabProjectMirrorSetToState(d *schema.ResourceData, projectMirror *gitlab.ProjectMirror, projectID *string) {
+
+	mirrorID := strconv.Itoa(projectMirror.ID)
+	d.Set("enabled", projectMirror.Enabled)
+	d.Set("only_protected_branches", projectMirror.OnlyProtectedBranches)
+	d.Set("keep_divergent_refs", projectMirror.KeepDivergentRefs)
+	d.SetId(buildTwoPartID(projectID, &mirrorID))
```

Comment:
Nitpick, but this line is unnecessary since in all cases the ID will have already been set before calling this function.